pallets/merkle-airdrop/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+        /// * `InvalidProof` - If the provided Merkle proof is invalid
+        /// * `InsufficientAirdropBalance` - If the airdrop doesn't have enough tokens
+        #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::claim())]
+        pub fn claim(
+            origin: OriginFor<T>,
+            airdrop_id: AirdropId,
+            recipient: T::AccountId,
+            amount: <<T as Config>::Currency as Currency<T::AccountId>>::Balance,
+            merkle_proof: Vec<[u8; 32]>,
+        ) -> DispatchResult {
+            // Ensure the call has no origin (can be called by anyone)
+            ensure_none(origin)?;
+
+            // Ensure the airdrop exists
+            ensure!(
+                AirdropMerkleRoots::<T>::contains_key(airdrop_id),
+                Error::<T>::AirdropNotFound
+            );
+
+            // Ensure the recipient hasn't already claimed
+            ensure!(
+                !Claimed::<T>::contains_key(airdrop_id, &recipient),
+                Error::<T>::AlreadyClaimed
+            );
+
+            // Get the Merkle root for this airdrop
+            let merkle_root =
+                AirdropMerkleRoots::<T>::get(airdrop_id).ok_or(Error::<T>::AirdropNotFound)?;
+
+            // Verify the Merkle proof using sender
+            ensure!(
+                Self::verify_merkle_proof(&recipient, amount, &merkle_root, &merkle_proof),
+                Error::<T>::InvalidProof
+            );
+
+            // Ensure the airdrop has sufficient balance
+            let airdrop_balance = AirdropBalances::<T>::get(airdrop_id)
+                .ok_or(Error::<T>::InsufficientAirdropBalance)?;
+            ensure!(
+                airdrop_balance >= amount,
+                Error::<T>::InsufficientAirdropBalance
+            );
+
+            // Mark as claimed before performing the transfer
+            Claimed::<T>::insert(airdrop_id, &recipient, true);
+
+            // Update the airdrop balance
+            AirdropBalances::<T>::mutate(airdrop_id, |balance| {
+                if let Some(current_balance) = balance {
+                    *current_balance = current_balance.saturating_sub(amount);
+                }
+            });
+
+            // Transfer tokens from the pallet account to the recipient
+            T::Currency::transfer(
+                &Self::account_id(),
+                &recipient,
+                amount,
+                frame_support::traits::ExistenceRequirement::KeepAlive,
+            )?;
+
+            // Emit an event using recipient
+            Self::deposit_event(Event::Claimed {
+                airdrop_id,
+                account: recipient,
+                amount,
+            });
+
+            Ok(())
+        }
+    }
+
+    #[pallet::validate_unsigned]
+    impl<T: Config> ValidateUnsigned for Pallet<T> {
+        type Call = Call<T>;
+
+        fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
```

Comment:
Is this so clients can check validity before pulling the trigger?

Reply:
Removed now :)